### PR TITLE
Fix for adding more authentication fields

### PR DESCRIPTION
**Pull request by JSeam2**
This pull request is in reference to 
#84 and #87 

The modification in #87 did not result in `JSONObtainWebToken` reflecting the added authentication fields. The fix involved referencing `jwt_settings.JWT_GET_USER_BY_NATURAL_KEY_HANDLER(username)` in `decorators.py` as well as in `backend.py` instead of using `utils.get_user_by_natural_key`.

After including the following code below in settings.py of the Django web. It is possible to make a graphql mutation using the additional fields `email`, `phone` by adding the respective strings into the `username` field or from the customized `get_user_model().USERNAME_FIELD` to log in. There is a need to ensure that `email`, `phone` are unique in the user model used.
```
from django.contrib.auth import get_user_model
from django.db.models import Q


def custom_get_user_by_natural_key(username):
    User = get_user_model()
    try:
        return User.objects.get(
            Q(username=username) | Q(email=username) | Q(phone=username)
        )
    except User.DoesNotExist:
        return None


GRAPHQL_JWT = {
    'JWT_GET_USER_BY_NATURAL_KEY_HANDLER': custom_get_user_by_natural_key
}
```

Example graphql query possible now
```
# using username
mutation {
  JSONObtainWebToken(username: "myUsername" password: "examplepassword") {
    token
    refreshToken
    
  }
}

# using email
mutation {
  JSONObtainWebToken(username: "myemail@example.com" password: "examplepassword") {
    token
    refreshToken
    
  }
}

# using phone number
mutation {
  JSONObtainWebToken(username: "12346789" password: "examplepassword") {
    token
    refreshToken
    
  }
}
```

### CODE REVIEW

File: graphql_jwt/backends.py [hunk 1]
```diff
@@ -16,4 +17,4 @@ def authenticate(self, request=None, **kwargs):
         return None
 
     def get_user(self, user_id):
-        return get_user_by_natural_key(user_id)
+        return jwt_settings.JWT_GET_USER_BY_NATURAL_KEY_HANDLER(user_id)
```
Comment (mongkok) [inline graphql_jwt/backends.py]: get_user It's not important here.

File: graphql_jwt/decorators.py [hunk 2]
```diff
@@ -80,9 +81,11 @@ def on_resolve(values):
 
         username = kwargs.get(get_user_model().USERNAME_FIELD)
 
+        user = jwt_settings.JWT_GET_USER_BY_NATURAL_KEY_HANDLER(username)
+
         user = authenticate(
             request=context,
-            username=username,
+            username=user.username,
```
Comment (mongkok) [inline graphql_jwt/decorators.py]: User model may not have the `username` field

**Comment by mongkok** [comment]
Hey @JSeam2 ,
you don't need to modify the package, just create an authentication backend and use your `JWT_GET_USER_BY_NATURAL_KEY_HANDLER` function to retrieve the user.

See: https://docs.djangoproject.com/en/3.0/topics/auth/customizing/

I close the PR, if you have any questions please let me know.

**Comment by JSeam2** [comment]
Good day @mongkok , 
The issue with just using `JWT_GET_USER_BY_NATURAL_KEY_HANDLER` on my side is that customizing this doesn't allow me to log in with either `username` or `email` should I set 
```
def custom_get_user_by_natural_key(username):
    User = get_user_model()

    try:
        return User.objects.get(
            Q(username__iexact=username) | Q(email__iexact=username)
        )
    except User.DoesNotExist:
        return None


GRAPHQL_JWT = {
    ...
    'JWT_GET_USER_BY_NATURAL_KEY_HANDLER': custom_get_user_by_natural_key
}
```

On my side, I created an account with `username = asdf`, `email = asdf@asdf.com`, and `password = asdf`.

When I log in with only the username, it is ok. 
```
mutation {
  obtainToken(
    username: "asdf"
    password: "asdf"
  ) {
    token 
    refreshToken
  }
}
```

When I log in with only the email, I get a `Please, enter valid credentials` error.
```
mutation {
  obtainToken(
    username: "asdf@asdf.com"
    password: "asdf"
  ) {
    token 
    refreshToken
  }
}
```

Hence, I decided to invoke `jwt_settings.JWT_GET_USER_BY_NATURAL_KEY_HANDLER(username)` in `decorator.py`. Do you have any solutions regarding this?

**Comment by mongkok** [comment]
Hi @JSeam2,
I think the right approach is to create an authentication backend:
```py
from django.contrib.auth import get_user_model

UserModel = get_user_model()


class CustomBackend:

    def authenticate(self, request, username=None, password=None, **kwargs):
        if username is None or password is None:
            return None

        user = custom_get_user_by_natural_key(username)

        if user is not None and user.check_password(password) and user.is_active:
            return user
        return None

    def get_user(self, user_id):
        return None
```
...and add the backend to your AUTHENTICATION_BACKENDS:

```py
# settings.py
AUTHENTICATION_BACKENDS = [
    'graphql_jwt.backends.JSONWebTokenBackend',
    'path.to.CustomBackend',
    'django.contrib.auth.backends.ModelBackend',
]
```

**Comment by JSeam2** [comment]
@mongkok Noted, thanks for the advice!

**Pull request closed without merging**